File: codex/src/domain/entities/endereco.ts
import { DomainError } from "../errors/domain-error.js";

export type EnderecoProps = {
  cep: string;
  logradouro: string;
  numero: string;
  complemento?: string;
  cidade: string;
  estado: string;
};

export class Endereco {
  readonly cep: string;
  readonly logradouro: string;
  readonly numero: string;
  readonly complemento: string;
  readonly cidade: string;
  readonly estado: string;

  constructor(props: EnderecoProps) {
    if (!props.cep.trim()) throw new DomainError("CEP e obrigatorio.");
    if (!props.logradouro.trim()) throw new DomainError("Logradouro e obrigatorio.");
    if (!props.numero.trim()) throw new DomainError("Numero e obrigatorio.");
    if (!props.cidade.trim()) throw new DomainError("Cidade e obrigatoria.");
    if (!props.estado.trim()) throw new DomainError("Estado e obrigatorio.");

    this.cep = props.cep;
    this.logradouro = props.logradouro;
    this.numero = props.numero;
    this.complemento = props.complemento ?? "";
    this.cidade = props.cidade;
    this.estado = props.estado;
  }
}

File: claude/backend/src/main/java/com/ecommerce/domain/Endereco.java
package com.ecommerce.domain;

import jakarta.persistence.Column;
import jakarta.persistence.Embeddable;

/**
 * Endereco de entrega informado no checkout (RF17). Objeto de valor embutido no Pedido.
 */
@Embeddable
public class Endereco {

    @Column(name = "endereco_cep")
    private String cep;

    @Column(name = "endereco_logradouro")
    private String logradouro;

    @Column(name = "endereco_numero")
    private String numero;

    @Column(name = "endereco_complemento")
    private String complemento;

    @Column(name = "endereco_cidade")
    private String cidade;

    @Column(name = "endereco_estado")
    private String estado;

    protected Endereco() {
        // exigido pelo JPA
    }

    public Endereco(String cep, String logradouro, String numero, String complemento,
                    String cidade, String estado) {
        this.cep = cep;
        this.logradouro = logradouro;
        this.numero = numero;
        this.complemento = complemento;
        this.cidade = cidade;
        this.estado = estado;
    }

    public String getCep() {
        return cep;
    }

    public String getLogradouro() {
        return logradouro;
    }

    public String getNumero() {
        return numero;
    }

    public String getComplemento() {
        return complemento;
    }

    public String getCidade() {
        return cidade;
    }

    public String getEstado() {
        return estado;
    }
}
